rpcserver.go

```diff
@@ -582,20 +583,48 @@ func (r *rpcServer) ConnectPeer(ctx context.Context,
 		return nil, fmt.Errorf("cannot make connection to self")
 	}
 
-	// If the address doesn't already have a port, we'll assume the current
-	// default port.
-	var addr string
-	_, _, err = net.SplitHostPort(in.Addr.Host)
+	var host net.Addr
+	addrLen := len(in.Addr.Host)
+	h, p, err := net.SplitHostPort(in.Addr.Host)
 	if err != nil {
-		addr = net.JoinHostPort(in.Addr.Host, strconv.Itoa(defaultPeerPort))
+		if (addrLen == 22 || addrLen == 62) && in.Addr.Host[addrLen-6:] == ".onion" {
```

Comment:
I also this this can be made more elegant like

```go
h, p, err := net.SplitHostPort(in.Addr.Host) 
if err != nil {
        hostStr = in.Addr.Host
        if isOnion(hostStr) {
               port = defaultOnionPort
        } else {
               port = defaultPeerPort
        }
} else {
       hostStr = h
       port = strconv(p)
}

if isOnion(hostStr) {
        host = &torsvc.OnionAddress { ...
} else {
        host, err = cfg.net.ResolveTCPAddr("tcp", hostStr)
}
```
with some comments sprinkled in there

Reply:
Or maybe 
```go
h, p, err := net.SplitHostPort(in.Addr.Host) 
if err != nil {
        hostStr = in.Addr.Host
        port = 0
} else {
       hostStr = h
       port = strconv(p)
}

if isOnion(hostStr) {
        if port == 0 {
               port = deafultOnionPort
        }
        host = &torsvc.OnionAddress { ...
} else {
        if port == 0 {
               port = deafultPeerPort
        }
        host, err = cfg.net.ResolveTCPAddr("tcp", hostStr)
}
```